File: _includes/sort-benchmark/algorithms/binomial_heap.rs
struct BinomialNode {
    key: usize,
    degree: u32,
    child: Option<Box<BinomialNode>>,
    sibling: Option<Box<BinomialNode>>,
}

fn link(mut child: Box<BinomialNode>, mut parent: Box<BinomialNode>) -> Box<BinomialNode> {
    child.sibling = parent.child.take();
    parent.child = Some(child);
    parent.degree += 1;
    parent
}

fn roots_to_vec(mut head: Option<Box<BinomialNode>>) -> Vec<Box<BinomialNode>> {
    let mut roots = Vec::new();
    while let Some(mut node) = head.take() {
        head = node.sibling.take();
        roots.push(node);
    }
    roots
}

fn vec_to_roots(roots: Vec<Box<BinomialNode>>) -> Option<Box<BinomialNode>> {
    let mut head: Option<Box<BinomialNode>> = None;
    let mut tail = &mut head;
    for node in roots {
        *tail = Some(node);
        tail = &mut tail.as_mut().unwrap().sibling;
    }
    head
}

fn merge_root_lists(
    mut a: Option<Box<BinomialNode>>,
    mut b: Option<Box<BinomialNode>>,
) -> Option<Box<BinomialNode>> {
    let mut merged = Vec::new();
    while a.is_some() && b.is_some() {
        if a.as_ref().unwrap().degree <= b.as_ref().unwrap().degree {
            let mut node = a.take().unwrap();
            a = node.sibling.take();
            merged.push(node);
        } else {
            let mut node = b.take().unwrap();
            b = node.sibling.take();
            merged.push(node);
        }
    }
    while let Some(mut node) = a.take() {
        a = node.sibling.take();
        merged.push(node);
    }
    while let Some(mut node) = b.take() {
        b = node.sibling.take();
        merged.push(node);
    }
    vec_to_roots(merged)
}

fn consolidate(head: Option<Box<BinomialNode>>) -> Option<Box<BinomialNode>> {
    let mut roots = roots_to_vec(head);
    let mut i = 0;
    while i + 1 < roots.len() {
        if roots[i].degree != roots[i + 1].degree {
            i += 1;
            continue;
        }
        // Three equal degrees: leave the first and merge the latter two (CLRS).
        if i + 2 < roots.len() && roots[i + 2].degree == roots[i].degree {
            i += 1;
            continue;
        }
        let a = roots.remove(i);
        let b = roots.remove(i);
        let linked = if a.key <= b.key {
            link(b, a)
        } else {
            link(a, b)
        };
        roots.insert(i, linked);
    }
    vec_to_roots(roots)
}

fn union(
    h1: Option<Box<BinomialNode>>,
    h2: Option<Box<BinomialNode>>,
) -> Option<Box<BinomialNode>> {
    consolidate(merge_root_lists(h1, h2))
}

fn insert_key(heap: Option<Box<BinomialNode>>, key: usize) -> Option<Box<BinomialNode>> {
    let node = Box::new(BinomialNode {
        key,
        degree: 0,
        child: None,
        sibling: None,
    });
    union(heap, Some(node))
}

fn reverse_children(mut child: Option<Box<BinomialNode>>) -> Option<Box<BinomialNode>> {
    let mut rev: Option<Box<BinomialNode>> = None;
    while let Some(mut node) = child.take() {
        child = node.sibling.take();
        node.sibling = rev;
        rev = Some(node);
    }
    rev
}

fn extract_min(heap: Option<Box<BinomialNode>>) -> (Option<usize>, Option<Box<BinomialNode>>) {
    let Some(head) = heap else {
        return (None, None);
    };

    let mut roots = roots_to_vec(Some(head));
    let mut min_i = 0;
    for i in 1..roots.len() {
        if roots[i].key < roots[min_i].key {
            min_i = i;
        }
    }

    let mut min_node = roots.remove(min_i);
    let key = min_node.key;
    let children = reverse_children(min_node.child.take());
    (Some(key), union(vec_to_roots(roots), children))
}

fn binomial_heap_sort(a: &mut [usize]) {
    let mut heap = None;
    for &key in a.iter() {
        heap = insert_key(heap, key);
    }
    for slot in a.iter_mut() {
        let (key, next) = extract_min(heap);
        heap = next;
        *slot = key.expect("binomial heap exhausted early");
    }
}

File: _includes/sort-benchmark/algorithms/binomial_heap.swift
final class BinomialNode {
    var key: Int
    var degree: UInt32
    var child: BinomialNode?
    var sibling: BinomialNode?

    init(key: Int, degree: UInt32 = 0, child: BinomialNode? = nil, sibling: BinomialNode? = nil) {
        self.key = key
        self.degree = degree
        self.child = child
        self.sibling = sibling
    }
}

func link(_ child: BinomialNode, _ parent: BinomialNode) -> BinomialNode {
    child.sibling = parent.child
    parent.child = child
    parent.degree += 1
    return parent
}

func roots_to_vec(_ head: BinomialNode?) -> [BinomialNode] {
    var head = head
    var roots = [BinomialNode]()
    while let node = head {
        head = node.sibling
        node.sibling = nil
        roots.append(node)
    }
    return roots
}

func vec_to_roots(_ roots: [BinomialNode]) -> BinomialNode? {
    var head: BinomialNode? = nil
    var tail: BinomialNode? = nil
    for node in roots {
        node.sibling = nil
        if head == nil {
            head = node
            tail = node
        } else {
            tail!.sibling = node
            tail = node
        }
    }
    return head
}

func merge_root_lists(_ a: BinomialNode?, _ b: BinomialNode?) -> BinomialNode? {
    var a = a
    var b = b
    var merged = [BinomialNode]()
    while let aNode = a, let bNode = b {
        if aNode.degree <= bNode.degree {
            a = aNode.sibling
            aNode.sibling = nil
            merged.append(aNode)
        } else {
            b = bNode.sibling
            bNode.sibling = nil
            merged.append(bNode)
        }
    }
    while let node = a {
        a = node.sibling
        node.sibling = nil
        merged.append(node)
    }
    while let node = b {
        b = node.sibling
        node.sibling = nil
        merged.append(node)
    }
    return vec_to_roots(merged)
}

func consolidate(_ head: BinomialNode?) -> BinomialNode? {
    var roots = roots_to_vec(head)
    var i = 0
    while i + 1 < roots.count {
        if roots[i].degree != roots[i + 1].degree {
            i += 1
            continue
        }
        // Three equal degrees: leave the first and merge the latter two (CLRS).
        if i + 2 < roots.count && roots[i + 2].degree == roots[i].degree {
            i += 1
            continue
        }
        let a = roots.remove(at: i)
        let b = roots.remove(at: i)
        let linked: BinomialNode
        if a.key <= b.key {
            linked = link(b, a)
        } else {
            linked = link(a, b)
        }
        roots.insert(linked, at: i)
    }
    return vec_to_roots(roots)
}

func union(_ h1: BinomialNode?, _ h2: BinomialNode?) -> BinomialNode? {
    consolidate(merge_root_lists(h1, h2))
}

func insert_key(_ heap: BinomialNode?, _ key: Int) -> BinomialNode? {
    let node = BinomialNode(key: key)
    return union(heap, node)
}

func reverse_children(_ child: BinomialNode?) -> BinomialNode? {
    var child = child
    var rev: BinomialNode? = nil
    while let node = child {
        child = node.sibling
        node.sibling = rev
        rev = node
    }
    return rev
}

func extract_min(_ heap: BinomialNode?) -> (Int?, BinomialNode?) {
    guard let head = heap else {
        return (nil, nil)
    }

    var roots = roots_to_vec(head)
    var min_i = 0
    for i in 1..<roots.count {
        if roots[i].key < roots[min_i].key {
            min_i = i
        }
    }

    let min_node = roots.remove(at: min_i)
    let key = min_node.key
    let children = reverse_children(min_node.child)
    min_node.child = nil
    return (key, union(vec_to_roots(roots), children))
}

func binomial_heap_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { binomial_heap_sort($0) }
}

func binomial_heap_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    var heap: BinomialNode? = nil
    for i in 0..<a.count {
        heap = insert_key(heap, a[i])
    }
    for i in 0..<a.count {
        let (key, next) = extract_min(heap)
        heap = next
        guard let key else {
            fatalError("binomial heap exhausted early")
        }
        a[i] = key
    }
}
